### PR TITLE
Callback with reload

### PR DESCRIPTION
**Pull request by rasapala**
### 🛠 Summary

JIRA CVS-151523
Current Async implementation does not handle online modifications - we need to move unlaod guard into callback in model instance async implementation.
### 🧪 Checklist

- [x] Unit tests added.
- [ ] The documentation updated.
- [x] Change follows security best practices.
``

### CODE REVIEW

File: src/test/stress_test_utils.hpp [hunk 1]
```diff
@@ -1881,7 +1908,93 @@ class ConfigChangeStressTest : public TestWithTempDir {
...
+            CallbackUnblockingStruct callbackStruct;
+            auto unblockSignal = callbackStruct.signal.get_future();
+            callbackStruct.response = response;
+
+            ASSERT_CAPI_STATUS_NULL(OVMS_InferenceRequestSetCompletionCallback(request, callbackUnblockingAndFreeingRequest, reinterpret_cast<void*>(&callbackStruct)));
+
+            OVMS_Status* status = OVMS_InferenceAsync(this->cserver, request);
+            //OVMS_InferenceRequestDelete(request);
+            // check
+            auto callbackReturnValue = unblockSignal.get();
```
Comment (atobiszei) [inline src/test/stress_test_utils.hpp:1980]: Here you should wait for signal only if InferenceAsync was successful

**Comment by atobiszei** [comment]
> JIRA CVS-151523 Current Async implementation does not handle online modifications - we need to move unlaod guard into 
unlaod -> unload

File: src/modelinstance.cpp [hunk 2]
```diff
@@ -1461,41 +1461,51 @@ Status ModelInstance::inferAsync(const RequestType* requestProto,
...
-    // TODO unload model guard & test
-    inferRequest.set_callback([this, requestProto, &inferRequest, userCallback, userCallbackData](std::exception_ptr exception) {
-        SPDLOG_INFO("Entry of ov::InferRequest callback call");
-        if (exception) {
+    {
+        inferRequest.set_callback([this, requestProto, &inferRequest, userCallback, userCallbackData, modelUnloadGuardPtrMoved = std::shared_ptr<ModelInstanceUnloadGuard>(std::move(modelUnloadGuardPtr))](std::exception_ptr exception) mutable {
+            SPDLOG_INFO("Entry of ov::InferRequest callback call");
+            if (exception) {
+                try {
+                    SPDLOG_INFO("rethrow_exception");
```
Comment (atobiszei) [inline src/modelinstance.cpp:1469]: Debug

File: src/modelinstance.cpp [hunk 3]
```diff
@@ -1461,41 +1461,51 @@ Status ModelInstance::inferAsync(const RequestType* requestProto,
     OVMS_InferenceRequestCompletionCallback_t userCallback = requestProto->getResponseCompleteCallback();
     void* userCallbackData = requestProto->getResponseCompleteCallbackData();
     // here pass by copy into callback
-    // TODO unload model guard & test
-    inferRequest.set_callback([this, requestProto, &inferRequest, userCallback, userCallbackData](std::exception_ptr exception) {
-        SPDLOG_INFO("Entry of ov::InferRequest callback call");
-        if (exception) {
+    {
+        inferRequest.set_callback([this, requestProto, &inferRequest, userCallback, userCallbackData, modelUnloadGuardPtrMoved = std::shared_ptr<ModelInstanceUnloadGuard>(std::move(modelUnloadGuardPtr))](std::exception_ptr exception) mutable {
+            SPDLOG_INFO("Entry of ov::InferRequest callback call");
```
Comment (atobiszei) [inline src/modelinstance.cpp:1466]: Debug

File: src/test/c_api_stress_tests.cpp [hunk 4]
```diff
@@ -64,6 +66,86 @@ class StressModelCapiConfigChanges : public StressCapiConfigChanges {
     }
 };
 
+TEST_F(ConfigChangeStressTestSingleModel, ChangeToEmptyConfigInference) {
+    bool performWholeConfigReload = true;  // we just need to have all model versions rechecked
+    std::set<StatusCode> requiredLoadResults = {
+        StatusCode::OK};  // we expect full continuity of operation
+    std::set<StatusCode> allowedLoadResults = {
```
Comment (atobiszei) [inline src/test/c_api_stress_tests.cpp:73]: Since end result of test is that model is unloaded then MODEL_VERSION_NOT_LOADED_ANYMORE is required state.

Comment (atobiszei) [reply src/test/c_api_stress_tests.cpp:73]: The same applies to test below.

File: src/test/c_api_stress_tests.cpp [hunk 5]
```diff
@@ -64,6 +66,86 @@ class StressModelCapiConfigChanges : public StressCapiConfigChanges {
...
+        allowedLoadResults);
+}
+
+TEST_F(ConfigChangeStressTestAsync, ChangeToEmptyConfigAsyncInference) {
+    bool performWholeConfigReload = true;  // we just need to have all model versions rechecked
+    std::set<StatusCode> requiredLoadResults = {
+        StatusCode::OK};  // we expect full continuity of operation
+    std::set<StatusCode> allowedLoadResults = {
+        StatusCode::MODEL_VERSION_NOT_LOADED_ANYMORE,
+        StatusCode::OV_INTERNAL_DESERIALIZATION_ERROR  // this should be hit if test is stressing enough, sporadically does not happen
```
Comment (atobiszei) [inline src/test/c_api_stress_tests.cpp:90]: This shouldn;t be hit now:
StatusCode::OV_INTERNAL_DESERIALIZATION_ERROR

File: src/test/c_api_stress_tests.cpp [hunk 6]
```diff
@@ -275,7 +357,7 @@ TEST_F(StressCapiConfigChanges, AddModelDuringGetModelStatusLoad) {
 
 class StressPipelineCustomNodesWithPreallocatedBuffersCapiConfigChanges : public StressCapiConfigChanges {
 public:
-    void checkInferResponse(OVMS_InferenceResponse* response) override {
+    void checkInferResponse(OVMS_InferenceResponse* response, std::string& expectedOutputName) override {
```
Comment (atobiszei) [inline src/test/c_api_stress_tests.cpp:360]: Not used?

Comment (rasapala) [reply src/test/c_api_stress_tests.cpp:360]: Yes, but it is needed to override and proper virtual call.

File: src/deserialization.hpp [hunk 7]
```diff
@@ -388,7 +388,9 @@ std::tuple<ovms::Status, const typename RequestTraits<ovms::InferenceRequest>::T
     }
     }
     if ((!status.ok() || requestTensorPtr == nullptr)) {
-        SPDLOG_DEBUG("Failed to deserialize request. Validation of request failed");
+        if (extractChoice == ExtractChoice::EXTRACT_INPUT) {
```
Comment (dkalinowski) [inline src/deserialization.hpp:391]: Why we dont need to log errors when choice is output?

Comment (atobiszei) [reply src/deserialization.hpp:391]: As long as we iterate over output map we try to get each output. It fails for requests with no output specified so it generated this error message for most of cases. If output is not specified its extraction will failed. We could leave this message if we would put filtered map of outputs info when doing deserialization.

File: src/modelinstance.cpp [hunk 8]
```diff
@@ -1461,41 +1461,51 @@ Status ModelInstance::inferAsync(const RequestType* requestProto,
...
+            if (exception) {
+                try {
+                    SPDLOG_DEBUG("rethrow_exception");
+                    std::rethrow_exception(exception);
+                } catch (const std::exception& e) {
+                    SPDLOG_DEBUG("got exception in ov::InferRequest callback: {}", e.what());
+                } catch (...) {
+                    SPDLOG_DEBUG("got exception in ov::InferRequest callback");
+                    return;
+                }
```
Comment (dkalinowski) [inline src/modelinstance.cpp:1476]: Is it possible to just simply use `exception` object as is? Why do we need to re-throw?

File: src/modelinstance.cpp [hunk 9]
```diff
@@ -1461,41 +1461,51 @@ Status ModelInstance::inferAsync(const RequestType* requestProto,
...
-        SPDLOG_DEBUG("Calling user provided callback");  // TODO check if this shows
-        userCallback(response, 1, userCallbackData);
-        SPDLOG_DEBUG("Called user provided callback");                       // TODO check if this shows
-        inferRequest.set_callback([](std::exception_ptr exception_ptr) {});  // reset callback on infer request // TODO this should be called on all exit paths
-    });
-    OV_LOGGER("ov::InferRequest: {}, inferRequest.start_async()", reinterpret_cast<void*>(&inferRequest));
-    inferRequest.start_async();  // TODO wrap into try catch
+            OVMS_InferenceResponse* response = reinterpret_cast<OVMS_InferenceResponse*>(res.release());
+            SPDLOG_DEBUG("Calling user provided callback");  // TODO check if this shows
+            userCallback(response, 1, userCallbackData);
```
Comment (dkalinowski) [inline src/modelinstance.cpp:1493]: What if user did not specify callback?

File: src/modelinstance.cpp [hunk 10]
```diff
@@ -1461,41 +1461,51 @@ Status ModelInstance::inferAsync(const RequestType* requestProto,
...
+            SPDLOG_DEBUG("Called user provided callback");                       // TODO check if this shows
+            inferRequest.set_callback([](std::exception_ptr exception_ptr) {});  // reset callback on infer request // TODO this should be called on all exit paths
```
Comment (dkalinowski) [inline src/modelinstance.cpp:1495]: is there task for all those todo? those seem critical in terms of release

File: src/test/stress_test_utils.hpp [hunk 11]
```diff
@@ -1879,9 +1929,175 @@ class ConfigChangeStressTest : public TestWithTempDir {
...
+            ASSERT_CAPI_STATUS_NULL(OVMS_InferenceRequestSetCompletionCallback(request, callbackUnblockingAndFreeingRequest, reinterpret_cast<void*>(&callbackStruct)));
+
+            OVMS_Status* status = OVMS_InferenceAsync(this->cserver, request);
+
+            uint32_t code = 0;
+            OVMS_Status* codeStatus = OVMS_StatusCode(status, &code);
+            StatusCode sc;
+            if (codeStatus != nullptr) {
+                sc = static_cast<StatusCode>(StatusCode::OK);
+            } else {
```
Comment (dkalinowski) [inline src/test/stress_test_utils.hpp:2009]: Used multiple times in multiple tests, maybe add helper?

File: src/modelinstance.cpp (hunk 8 above)
Comment (rasapala) [reply src/modelinstance.cpp:1476]: I think this was used as BKM example when using callback in OV.

File: src/modelinstance.cpp (hunk 9 above)
Comment (rasapala) [reply src/modelinstance.cpp:1493]: I think we will log errors and stop the async execution. This will be done in error handling task.

Comment (dkalinowski) [reply src/modelinstance.cpp:1493]: I think we should at least check if the callback is not null to be safe here

File: src/modelinstance.cpp (hunk 10 above)
Comment (rasapala) [reply src/modelinstance.cpp:1495]: We have additional tasks in epic. Todos will be reviewed.

File: src/test/c_api_stress_tests.cpp [hunk 12]
```diff
@@ -64,6 +66,78 @@ class StressModelCapiConfigChanges : public StressCapiConfigChanges {
...
+        requiredLoadResults,
+        allowedLoadResults);
+}
+
+TEST_F(ConfigChangeStressTestAsync, ChangeToWrongShapeAsyncInference) {
+    bool performWholeConfigReload = true;  // we just need to have all model versions rechecked
+    std::set<StatusCode> requiredLoadResults = {
+        StatusCode::OK};  // we expect full continuity of operation
+    std::set<StatusCode> allowedLoadResults = {
+        StatusCode::INVALID_SHAPE};
```
Comment (atobiszei) [inline src/test/c_api_stress_tests.cpp:102]: Isn't this code expected end status - hence shouldn't it be in required list?

**Pull request closed without merging**